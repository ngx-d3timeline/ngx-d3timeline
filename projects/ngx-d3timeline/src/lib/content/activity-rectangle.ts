import { identifier } from '../types';

export interface ActivityRectangle {
  id: identifier;
  title: string;
  showTitle: boolean;
  transform: string;
  width: number;
  height: number;
  fontSize: number;
}
