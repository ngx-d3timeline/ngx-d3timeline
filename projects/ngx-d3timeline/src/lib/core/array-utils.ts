export function sum(...args: number[]) {
  return args.reduce((a, b) => a + b);
}
