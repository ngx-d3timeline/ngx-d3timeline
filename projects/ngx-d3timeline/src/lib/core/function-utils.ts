export function partialApply<TInput, TAppliedArgs extends any[], TOut>(
  func: (input: TInput, ...appliedArgs: TAppliedArgs) => TOut
): (...appliedArgs: TAppliedArgs) => (input: TInput) => TOut {
  return (...appliedArgs: TAppliedArgs) => (input: TInput) =>
    func(input, ...appliedArgs);
}

export function pipe<T>(
  ...args: ((...args: any[]) => T)[]
): (...t: any[]) => T {
  return x => args.reduce((acc, curr) => curr(acc), x);
}

export function clampZero(x: number) {
  return Math.max(x, 0);
}
