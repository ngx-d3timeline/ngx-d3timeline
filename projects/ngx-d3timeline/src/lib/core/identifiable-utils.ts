import { Identifiable, identifier } from './types';

export function findIdentifiable<T extends Identifiable>(
  identifiables: T[],
  id: identifier
): T {
  return id && identifiables.find(identifiable => identifiable.id === id);
}
