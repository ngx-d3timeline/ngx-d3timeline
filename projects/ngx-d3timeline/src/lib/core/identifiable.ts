export type Identifier = string | number;

export interface Identifiable {
  id: Identifier;
}

export function findIdentifiable<T extends Identifiable>(
  identifiables: T[],
  id: Identifier
): T {
  return id && identifiables.find(identifiable => identifiable.id === id);
}
