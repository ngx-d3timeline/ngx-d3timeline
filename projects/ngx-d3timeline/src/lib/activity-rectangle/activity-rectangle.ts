import { identifier } from '../core/types';

export interface ActivityRectangle {
  id: identifier;
  title: string;
  showTitle: boolean;
  transform: string;
  width: number;
  height: number;
  fontSize: number;
}
