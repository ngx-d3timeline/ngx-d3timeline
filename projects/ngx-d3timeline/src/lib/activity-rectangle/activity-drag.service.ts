import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { Store } from '../store-lib/store';
import { withLatestFrom, filter, map } from 'rxjs/operators';
import { drag } from 'd3-drag';
import { select, event } from 'd3-selection';

import * as fromActions from '../store/actions';
import { Identifier } from '../core/identifiable';
import { selectActivityUpdatedForDrag } from '../activity/activity.selectors';

@Injectable({ providedIn: 'root' })
export class ActivityDragService {
  private dragEndSubject = new BehaviorSubject(null);

  private activityUpdatedForDrag$ = this.dragEndSubject.pipe(
    filter(() => !!event),
    withLatestFrom(this.store.select(selectActivityUpdatedForDrag)),
    map(([, draggedToActivity]) => draggedToActivity)
  );

  constructor(private store: Store) {
    this.activityUpdatedForDrag$.subscribe(activityUpdatedForDrag => {
      this.store.dispatch(
        new fromActions.TimelineDragEndedAction(activityUpdatedForDrag)
      );
    });
  }

  setupDrag(id: Identifier, nativeElement: HTMLElement) {
    const onDrag = drag()
      .on('drag', () => this.onDragging(id))
      .on('end', this.onDragEnded.bind(this));

    onDrag(select(nativeElement));
  }

  private onDragging(id: Identifier) {
    this.store.dispatch(new fromActions.TimelineDraggingAction({ id, event }));
  }

  private onDragEnded() {
    this.dragEndSubject.next(null);
  }
}
