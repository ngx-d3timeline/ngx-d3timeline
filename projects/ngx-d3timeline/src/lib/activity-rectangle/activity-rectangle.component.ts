import {
  Component,
  Input,
  ElementRef,
  ViewChild,
  AfterViewInit,
  ChangeDetectionStrategy
} from '@angular/core';
import { ActivityRectangle } from './activity-rectangle';
import { ActivityRectangleService } from './activity-rectangle.service';
import { activityTitlePadding } from '../content/content-utils';

@Component({
  selector: '[ngx-d3timeline-activity-rectangle]',
  template: `
    <svg:g
      class="activity-rectangle"
      [attr.transform]="activityRectangle.transform"
      *ngIf="activityRectangle"
      #activityRectangleEl
    >
      <svg:rect
        [attr.height]="activityRectangle.height"
        [attr.width]="activityRectangle.width"
      ></svg:rect>
      <svg:g *ngIf="activityRectangle.showTitle">
        <svg:text
          dominant-baseline="hanging"
          [attr.dx]="activityTitlePadding"
          [attr.dy]="activityTitlePadding"
          [attr.font-size]="activityRectangle.fontSize"
        >
          {{ activityRectangle.title }}
        </svg:text>
      </svg:g>
    </svg:g>
  `,
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class ActivityRectangleComponent implements AfterViewInit {
  @Input() activityRectangle: ActivityRectangle;

  @ViewChild('activityRectangleEl') activityRectangleEl: ElementRef;

  activityTitlePadding = activityTitlePadding;

  constructor(public facade: ActivityRectangleService) {}

  ngAfterViewInit() {
    this.facade.setupDrag(this.activityRectangle, this.activityRectangleEl);
  }
}
