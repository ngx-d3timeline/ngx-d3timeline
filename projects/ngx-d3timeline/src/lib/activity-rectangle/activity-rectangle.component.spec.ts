import { TestBed, ComponentFixture } from '@angular/core/testing';
import { ActivityRectangleComponent } from './activity-rectangle.component';

describe('ActivityRectangleComponent', () => {
  let fixture: ComponentFixture<ActivityRectangleComponent>;

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [ActivityRectangleComponent]
    });

    fixture = TestBed.createComponent(ActivityRectangleComponent);
  });

  it('should not render if activityRectangle is null', () => {
    fixture.componentInstance.activityRectangle = null;

    fixture.detectChanges();

    expect(fixture.nativeElement).toMatchSnapshot();
  });

  it('should render correctly', () => {
    fixture.componentInstance.activityRectangle = {
      id: 1,
      title: 'Activity 1',
      type: 'Driving',
      transform: 'translate(50,0)',
      width: 50,
      height: 80,
      showTitle: true,
      fontSize: 10,
      fontFace: 'Arial'
    };

    fixture.detectChanges();

    expect(fixture.nativeElement).toMatchSnapshot();
  });

  it('should hide title when showTitle false', () => {
    fixture.componentInstance.activityRectangle = {
      id: 1,
      title: 'Activity 1',
      transform: 'translate(50,0)',
      type: 'Driving',
      width: 50,
      height: 80,
      showTitle: false,
      fontSize: 10,
      fontFace: 'Arial'
    };

    fixture.detectChanges();

    expect(fixture.nativeElement).toMatchSnapshot();
  });
});
