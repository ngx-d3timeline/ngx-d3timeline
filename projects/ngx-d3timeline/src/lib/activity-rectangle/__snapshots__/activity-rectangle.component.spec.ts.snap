// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActivityRectangleComponent should hide title when showTitle false 1`] = `
<div
  id="root2"
  ngx-d3timeline-activity-rectangle=""
>
  
  <g
    class="activity-rectangle"
    transform="translate(50,0)"
  >
    <rect
      height="80"
      width="50"
    />
    
  </g>
</div>
`;

exports[`ActivityRectangleComponent should not render if activityRectangle is null 1`] = `
<div
  id="root0"
  ngx-d3timeline-activity-rectangle=""
>
  
</div>
`;

exports[`ActivityRectangleComponent should render correctly 1`] = `
<div
  id="root1"
  ngx-d3timeline-activity-rectangle=""
>
  
  <g
    class="activity-rectangle"
    transform="translate(50,0)"
  >
    <rect
      height="80"
      width="50"
    />
    
    <g>
      <text
        dominant-baseline="hanging"
        dx="2"
        dy="2"
        font-family="Arial"
        font-size="10"
      >
         Activity 1 
      </text>
    </g>
  </g>
</div>
`;
